src/runtime-rs/crates/hypervisor/src/dragonball/inner.rs

```diff
@@ -0,0 +1,300 @@
+        nix::mount::umount2(path.as_str(), nix::mount::MntFlags::MNT_DETACH)
+            .with_context(|| format!("umount path {}", &path))
+    }
+
+    pub(crate) fn get_resource(&self, src: &str, dst: &str) -> Result<String> {
+        if self.jailed {
+            self.jail_resource(src, dst)
+        } else {
+            Ok(src.to_string())
+        }
+    }
+
+    fn jail_resource(&self, src: &str, dst: &str) -> Result<String> {
+        info!(sl!(), "src {} dst {}", src, dst);
+        if src.is_empty() || dst.is_empty() {
+            return Err(anyhow!("invalid param src {} dst {}", src, dst));
+        }
+
+        let jailed_location = [self.jailer_root.as_str(), dst].join("/");
+        mount::bind_mount_unchecked(src, jailed_location.as_str(), false).context("bind_mount")?;
+
+        let mut abs_path = String::from("/");
+        abs_path.push_str(dst);
+        Ok(abs_path)
+    }
+
+    fn set_boot_source(&mut self, kernel_path: &str, kernel_params: &str) -> Result<()> {
+        info!(
+            sl!(),
+            "kernel path {} kernel params {}", kernel_path, kernel_params
+        );
+
+        let mut boot_cfg = BootSourceConfig {
+            kernel_path: self
+                .get_resource(kernel_path, DRAGONBALL_KERNEL)
+                .context("jail resource")?,
+            ..Default::default()
+        };
+
+        if !kernel_params.is_empty() {
+            boot_cfg.boot_args = Some(kernel_params.to_string());
+        }
+
+        self.vmm_instance
+            .put_boot_source(boot_cfg)
+            .context("put boot source")
+    }
+
+    fn set_vm_rootfs(&mut self, path: &str, driver: &str) -> Result<()> {
+        info!(sl!(), "==== set vm rootfs {} {}", path, driver);
+        let jail_drive = self
+            .get_resource(path, DRAGONBALL_ROOT_FS)
+            .context("get resource")?;
+
+        if driver == VM_ROOTFS_DRIVER_BLK {
+            let blk_cfg = BlockDeviceConfigInfo {
+                path_on_host: PathBuf::from(jail_drive),
+                drive_id: DRAGONBALL_ROOT_FS.to_string(),
+                is_root_device: false,
+                // Add it as a regular block device
+                // This allows us to use a partitioned root block device
+                // is_read_only
+                is_read_only: true,
+                is_direct: false,
+                ..Default::default()
+            };
+
+            self.vmm_instance
+                .insert_block_device(blk_cfg)
+                .context("inert block device")
+        } else {
+            Err(anyhow!("Unknown vm_rootfs driver: {}", driver))
+        }
+    }
+
+    fn start_vmm_instance(&mut self) -> Result<()> {
+        info!(sl!(), "Starting VM");
+        self.vmm_instance
+            .instance_start()
+            .context("Failed to start vmm")?;
+        self.state = VmmState::VmRunning;
+        Ok(())
+    }
+
+    // wait_vmm will wait for timeout seconds for the VMM to be up and running.
```

Comment:
Nit: If `wait_vmm` is the fn name, then it should be `wait_vmm_ready`

Reply:
Thanks for your review, already updated.